executor/executor.go

```diff
@@ -5,175 +5,228 @@ package executor
+		for _, bt := range t.blocked {
+			// If we are the last dependency, mark the task as executable.
+			if bt.dependencies.Add(-1) > 0 {
+				continue
 			}
+			e.executable <- bt
 		}
+		t.blocked = nil // free memory
+		t.executed = true
+		t.l.Unlock()
+		e.outstanding.Done()
 	}()
+
+	// We avoid doing this check when adding tasks to the queue
+	// because it would require more synchronization.
+	if e.err.Load() != nil {
+		return
+	}
+
+	// Execute the task
+	if err := t.f(); err != nil {
+		e.err.CompareAndSwap(nil, err)
+		return
+	}
 }
 
 // Run executes [f] after all previously enqueued [f] with
-// overlapping [conflicts] are executed.
-// TODO: Handle read-only/write-only keys (currently the executor
-// treats everything still as ReadWrite, see issue below)
-// https://github.com/ava-labs/hypersdk/issues/709
-func (e *Executor) Run(conflicts state.Keys, f func() error) {
-	e.l.Lock()
-	defer e.l.Unlock()
+// overlapping [keys] are executed.
+//
+// Run is not safe to call concurrently.
+//
+// If there is an error, all remaining task execution will be skipped. It is up
+// to the caller to ensure correctness does not depend on exactly when work begins
+// to be skipped (i.e. not safe to rely on this functionality for block verification).
+func (e *Executor) Run(keys state.Keys, f func() error) {
+	e.outstanding.Add(1)
 
 	// Add task to map
-	id := len(e.tasks)
+	id := e.tasks
+	e.tasks++
 	t := &task{
-		id: id,
-		f:  f,
+		id:     id,
+		f:      f,
+		shared: []*task{},
+
+		blocked: make(map[int]*task),
+		readers: make(map[int]*task),
 	}
-	e.tasks[id] = t
+
+	// Add maximum number of dependencies to ensure we don't execute the task
+	// before we are finished enqueuing all dependencies.
+	//
+	// We can have more than 1 dependency per key (in the case that there
+	// are many readers for a single key), so we set this higher than we ever
+	// expect to see.
+	t.dependencies.Add(maxDependencies)
 
 	// Record dependencies
-	for k := range conflicts {
-		latest, ok := e.edges[k]
+	dependencies := set.NewSet[int](len(keys))
+	for k, v := range keys {
+		lt, ok := e.nodes[k]
 		if ok {
-			lt := e.tasks[latest]
-			if !lt.executed {
-				if t.dependencies == nil {
-					t.dependencies = set.NewSet[int](defaultSetSize)
-				}
-				t.dependencies.Add(lt.id)
-				if lt.blocking == nil {
-					lt.blocking = set.NewSet[int](defaultSetSize)
+			lt.l.Lock()
+			if v == state.Read {
+				// If we don't need exclusive access to a key, just mark
+				// that we are reading it and that we are a reader of it.
+				t.shared = append(t.shared, lt)
+				lt.readers[id] = t
+			} else {
+				// If we do need exclusive access to a key, we need to
+				// mark ourselves blocked on all readers ahead of us.
+				//
+				// If a task is a reader, that means it is not executed yet
+				// and can't mark itself as executed until all [shared] are
+				// cleared (which can't be done while we hold the lock for [lt]).
+				for _, rt := range lt.readers {
```

Comment:
I _think_ there might be a bug here, but I could be wrong. Consider two transactions, T1 and T2 both accessing key A and B. Suppose in the slow path, T1 and T2 have not executed.

T1 requests: A read, B write

T2 requests: A read, B write

In T1, these keys don't exist in `e.nodes` so we'll just add it

In T2, when we encounter A, `lt` is T1

We add ourselves, T2 to `lt.readers` for key A, and note that T1 hasn't executed so T1 is blocking T2. 

For key B in T2, we go through the `lt.readers` now since we're a Write, and one of the `rt` is us, T2. So we're saying T2 is blocked on ourself

Reply:
See `TestTwoConflictKeys` for a simplified unit test and see `TestLargeRandomReadsAndWrites` for arbitrary requesting